### PR TITLE
Graph Tessellation Benchmarks

### PR DESCRIPTION
**Pull request by QilongTang**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

This PR: 
1. Split the graph run earlier (which include evaluation, tessellation and run) into two part. I. Run II. Tessellation and Run
2. Moved GeometryHolder class to DynamoCore so test class can refer to easily. FYI: @sm6srw I think this is OK but let me know if you have concerns
image

### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner @scottmitchell 

### FYIs
@sm6srw

### CODE REVIEW

**Comment by mjkkirschner** [comment]
```
"C:\projects\dynamo\src\Dynamo.All.sln" (default target) (1) ->
7155"C:\projects\dynamo\test\DynamoCoreWpfTests\DynamoCoreWpfTests.csproj" (default target) (62) ->
7156(CoreCompile target) -> 
7157  GraphViewTests.cs(41,54): error CS0122: 'DefaultWatch3DViewModel.UnregisterNodeEventHandlers(NodeModel)' is inaccessible due to its protection level [C:\projects\dynamo\test\DynamoCoreWpfTests\DynamoCoreWpfTests.csproj] ```

File: test/DynamoCoreWpfTests/GraphViewTests.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+using Dynamo.Models;
+using Dynamo.Visualization;
+using Dynamo.Wpf.Rendering;
+using DynamoCoreWpfTests.Utility;
+
+namespace DynamoCoreWpfTests
+{
+
+    /// <summary>
+    /// This class should not be used for DynamoView unit tests
```
Comment (mjkkirschner) [inline test/DynamoCoreWpfTests/GraphViewTests.cs]: maybe add what it should be used for 😄 ?

Comment (QilongTang) [reply test/DynamoCoreWpfTests/GraphViewTests.cs]: yup, I can add

**Comment by QilongTang** [comment]
Running self CI and ready for review

File: test/DynamoCoreWpfTests/GraphViewTests.cs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        {
+            // A more precise way to measure the time spent is from when
+            // run() called to when even EvaluationCompleted is trigger on workspace
+            (Model.CurrentWorkspace as HomeWorkspaceModel).EvaluationCompleted += OnGraphEvaluationCompleted;
+
+            base.Run();
+            signalEvent.WaitOne();
+            signalEvent.Reset();
+            // Unsubscribe
+            (Model.CurrentWorkspace as HomeWorkspaceModel).EvaluationCompleted -= OnGraphEvaluationCompleted;
```
Comment (aparajit-pratap) [inline test/DynamoCoreWpfTests/GraphViewTests.cs]: Is it required to empty the Dispatcher queue here as well (by calling DispatcherUtils.DoEvents)?

File: src/DynamoCoreWpf/Properties/AssemblyInfo.cs [hunk 3]
```diff
@@ -35,3 +35,4 @@
 [assembly: InternalsVisibleTo("ViewExtensionLibraryTests")]
 [assembly: InternalsVisibleTo("DynamoWPFCLI")]
 [assembly: InternalsVisibleTo("CommandLineTests")]
+[assembly: InternalsVisibleTo("Watch3DNodeModelsWpf")]
```
Comment (aparajit-pratap) [inline src/DynamoCoreWpf/Properties/AssemblyInfo.cs]: Why is this needed? Are you exposing internals from `DynamoCoreWpf` in `Watch3DNodeModelsWpf`?

File: test/DynamoCoreWpfTests/GraphViewTests.cs [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        }
+
+        /// <summary>
+        /// Function used to measure graph geometry tessellation
+        /// </summary>
+        public void Tessellation()
```
Comment (aparajit-pratap) [inline test/DynamoCoreWpfTests/GraphViewTests.cs]: I would name this method with an action verb, like `Tessellate`, no?

File: src/DynamoCoreWpf/Properties/AssemblyInfo.cs (hunk 3 above)
Comment (QilongTang) [reply src/DynamoCoreWpf/Properties/AssemblyInfo.cs]: Yes, I changed `UnregisterNodeEventHandlers` from protected to internal so now this is needed because multiple other libraries need access after this PR

File: tools/Performance/DynamoPerformanceTests/DynamoViewPerformanceTestBase.cs [hunk 5]
```diff
@@ -69,15 +69,34 @@ public void IterationSetupOpenModelWithUI()
...
 
+            // open the dyn file
+            Open(DynamoFilePath);
+
+            // Disable tessellation or rendering
+            DisableRendering();
+        }
+
+        /// <summary>
+        /// Setup method to be called before each RenderGraph benchmark.
```
[the remark this answers is not in the document]
Comment (QilongTang) [reply tools/Performance/DynamoPerformanceTests/DynamoViewPerformanceTestBase.cs]: oh, will update

File: test/DynamoCoreWpfTests/GraphViewTests.cs [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+using Dynamo.Models;
+using Dynamo.Visualization;
+using Dynamo.Wpf.Rendering;
+using DynamoCoreWpfTests.Utility;
+
+namespace DynamoCoreWpfTests
+{
+
+    /// <summary>
+    /// This class should not be used for Dynamo View unit tests
```
[the remark this answers is not in the document]
Comment (QilongTang) [reply test/DynamoCoreWpfTests/GraphViewTests.cs]: hmm good point, the reason I leave it here is because I want to make sure this class always builds fine for us to use in performance benchmarks to run. Do you have a preference?

File: test/DynamoCoreWpfTests/GraphViewTests.cs (hunk 2 above)
Comment (QilongTang) [reply test/DynamoCoreWpfTests/GraphViewTests.cs]: I don't think so since we will not be checking preview bubble values, so no need to finish remaining tasks in the queue

File: test/DynamoCoreWpfTests/GraphViewTests.cs (hunk 4 above)
Comment (QilongTang) [reply test/DynamoCoreWpfTests/GraphViewTests.cs]: I can update this

**Comment by QilongTang** [comment]
Addressed some comments. Self CI passed, let me know what you think @aparajit-pratap

File: test/DynamoCoreWpfTests/GraphViewTests.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (aparajit-pratap) [reply test/DynamoCoreWpfTests/GraphViewTests.cs]: Yes, I would prefer just moving it to the performance project so there's less clutter in the main solution.

[the remark this answers is not in the document]
Comment (QilongTang) [reply test/DynamoCoreWpfTests/GraphViewTests.cs]: OK, This will add bunch of new dependencies if you are aware. I will update this PR shortly

**Comment by QilongTang** [comment]
All checks passed @aparajit-pratap
